File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Trapping Rain Water/two pointer/Solution.java
class Solution {
    public int trap(int[] height) {
        int result = 0;
        int left = 0;
        int right = height.length - 1;
        int leftMax = height[left];
        int rightMax = height[right];
        while (left < right) {
            if (leftMax <= rightMax) {
                left++;
                int currentHeight = height[left];
                if (currentHeight < leftMax) {
                    result += leftMax - currentHeight;
                } else {
                    leftMax = currentHeight;
                }
            } else {
                right--;
                int currentHeight = height[right];
                if (currentHeight < rightMax) {
                    result += rightMax - currentHeight;
                } else {
                    rightMax = currentHeight;
                }
            }
        }

        return result;
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Trapping Rain Water/two pointer/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        answer = 0
        new_height = height[:]
        left = 0
        left_maximum = height[left]
        right = len(height) - 1
        right_maximum = height[right]

        while left < right:
            if left_maximum <= right_maximum:
                left += 1
                if height[left] < left_maximum:
                    new_height[left] = left_maximum
                else:
                    left_maximum = height[left]
            else:
                right -= 1
                if height[right] < right_maximum:
                    new_height[right] = right_maximum
                else:
                    right_maximum = height[right]
        
        for i in range(len(height)):
            answer += new_height[i] - height[i]
        
        return answer
